core/src/main/java/org/jdbi/v3/ColonPrefixStatementRewriter.java

```diff
@@ -61,45 +56,71 @@
     @Override
     public RewrittenStatement rewrite(String sql, Binding params, StatementContext ctx)
     {
-        ParsedStatement stmt = cache.get(sql);
-        if (stmt == null) {
-            try {
-                stmt = parseString(sql);
-                cache.put(sql, stmt);
-            }
-            catch (IllegalArgumentException e) {
-                throw new UnableToCreateStatementException("Exception parsing for named parameter replacement", e, ctx);
+        try {
+            ParsedStatement stmt = cache.computeIfAbsent(rewriteDefines(sql, ctx), this::rewriteNamedParameters);
+            return new MyRewrittenStatement(stmt, ctx);
+        }
+        catch (IllegalArgumentException e) {
+            throw new UnableToCreateStatementException("Exception parsing for named parameter replacement", e, ctx);
+        }
+    }
+
+    String rewriteDefines(String sql, StatementContext ctx)
+    {
+        StringBuilder b = new StringBuilder();
+        DefineStatementLexer lexer = new DefineStatementLexer(new ANTLRStringStream(sql));
+        Token t = lexer.nextToken();
+        while (t.getType() != DefineStatementLexer.EOF) {
+            switch (t.getType()) {
+                case DefineStatementLexer.COMMENT:
+                case DefineStatementLexer.LITERAL:
+                case DefineStatementLexer.QUOTED_TEXT:
+                case DefineStatementLexer.DOUBLE_QUOTED_TEXT:
+                    b.append(t.getText());
+                    break;
+                case DefineStatementLexer.DEFINE:
+                    String text = t.getText();
+                    String key = text.substring(1, text.length() - 1);
+                    Object value = ctx.getAttribute(key);
+                    if (value == null) {
+                        throw new IllegalArgumentException("Undefined attribute for token '" + text + "'");
+                    }
+                    b.append(value);
+                    break;
+                case DefineStatementLexer.ESCAPED_TEXT:
+                    b.append(t.getText().substring(1));
+                    break;
+                default:
+                    break;
             }
+            t = lexer.nextToken();
         }
-        return new MyRewrittenStatement(stmt, ctx);
+        return b.toString();
```

Comment:
This same method is duplicated in ColonPrefixStatementRewriter and HashPrefixStatementRewriter. I had an idea of extracting it to e.g. `DefineStatementRewriter` and making it a wrapper around another `StatementRewriter` that it delegates to after rewriting the define tokens. Thoughts?